From 1889bc14fc998120bc10f963bc7814c472dfb2e2 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 5 Apr 2014 00:14:38 -0400
Subject: [PATCH] World.getNearbyEntities API

Adds much needed API for getting entities by location and not an entity base point.
---
 src/main/java/net/minecraft/server/Chunk.java      |  3 ++-
 .../java/org/bukkit/craftbukkit/CraftWorld.java    | 31 ++++++++++++++++++++++
 2 files changed, 33 insertions(+), 1 deletion(-)

diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index f11e978..987c9f3 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -860,7 +860,8 @@ public class Chunk {
             for (int l = 0; l < list1.size(); ++l) {
                 Entity entity = (Entity) list1.get(l);
 
-                if (oclass.isAssignableFrom(entity.getClass()) && entity.boundingBox.b(axisalignedbb) && (ientityselector == null || ientityselector.a(entity))) {
+                if ((oclass == null || oclass.isAssignableFrom(entity.getClass())) // EMC
+                    && entity.boundingBox.b(axisalignedbb) && (ientityselector == null || ientityselector.a(entity))) {
                     list.add(entity);
                 }
             }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 6b946ee..85544da 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -85,6 +85,37 @@ public class CraftWorld implements World {
         return getChunkAt(x >> 4, z >> 4).getBlock(x & 0xF, y & 0xFF, z & 0xF);
     }
 
+    // EMC start
+    public Collection<Entity> getNearbyEntities(Location loc, int x, int y, int z) {
+        return getNearbyEntitiesByType(null, loc, x, y, z);
+    }
+    public Collection<LivingEntity> getNearbyLivingEntities(Location loc, int x, int y, int z) {
+        return getNearbyEntitiesByType(org.bukkit.entity.LivingEntity.class, loc, x, y, z);
+    }
+    public Collection<Player> getNearbyPlayers(Location loc, int x, int y, int z) {
+        return getNearbyEntitiesByType(org.bukkit.entity.Player.class, loc, x, y, z);
+    }
+    public <T> Collection<T> getNearbyEntitiesByType(Class<? extends T> clazz, Location loc, int x, int y, int z) {
+        final double lx = loc.getX();
+        final double ly = loc.getY();
+        final double lz = loc.getZ();
+        final AxisAlignedBB aabb = AxisAlignedBB.a(lx, ly, lz, lx, ly, lz).a(x,y,z);
+
+        final List<net.minecraft.server.Entity> notchEntityList = ((CraftWorld) loc.getWorld()).getHandle().a(null, aabb, (IEntitySelector) null);
+        List<T> bukkitEntityList = new java.util.ArrayList<T>(notchEntityList.size());
+
+        for (net.minecraft.server.Entity e : notchEntityList) {
+            CraftEntity bukkitEntity = e.getBukkitEntity();
+            do {
+                if (clazz == null || clazz.isAssignableFrom(bukkitEntity.getClass())) {
+                    bukkitEntityList.add((T) bukkitEntity);
+                }
+            } while ((bukkitEntity = (CraftEntity) bukkitEntity.getPassenger()) != null);
+        }
+        return bukkitEntityList;
+    }
+    // EMC end
+
     public int getBlockTypeIdAt(int x, int y, int z) {
         return world.getTypeId(x, y, z);
     }
-- 
1.9.1

